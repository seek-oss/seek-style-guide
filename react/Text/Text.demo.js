import Text from './Text';
import * as sketch from './Text.sketch';

export default {
  route: '/text',
  title: 'Text',
  category: 'Typography',
  component: Text,
  sketch,
  initialProps: {
    children: 'Living Style Guide',
    hero: true
  },
  options: [
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Strong',
          transformProps: props => ({
            ...props,
            strong: true
          })
        },
        {
          label: 'Regular',
          transformProps: props => ({
            ...props,
            regular: true
          })
        },
        {
          label: 'Light',
          transformProps: props => ({
            ...props,
            light: true
          })
        },
        {
          label: 'Highlight',
          transformProps: props => ({
            ...props,
            highlight: true
          })
        },
        {
          label: 'Disable baseline',
          transformProps: props => ({
            ...props,
            baseline: false
          })
        }
      ]
    },
    {
      label: 'Tone',
      type: 'radio',
      states: [
        {
          label: 'Neutral tone',
          transformProps: props => props
        },
        {
          label: 'Positive',
          transformProps: props => ({
            ...props,
            positive: true
          })
        },
        {
          label: 'Critical',
          transformProps: props => ({
            ...props,
            critical: true
          })
        },
        {
          label: 'Secondary',
          transformProps: props => ({
            ...props,
            secondary: true
          })
        },
        {
          label: 'Info',
          transformProps: props => ({
            ...props,
            info: true
          })
        }
      ]
    },
    {
      label: 'Size',
      type: 'radio',
      states: [
        {
          label: 'Hero',
          transformProps: props => props
        },
        {
          label: 'Headline',
          transformProps: ({ hero, ...props }) => ({
            ...props,
            headline: true
          })
        },
        {
          label: 'Heading',
          transformProps: ({ hero, ...props }) => ({
            ...props,
            heading: true
          })
        },
        {
          label: 'Subheading',
          transformProps: ({ hero, ...props }) => ({
            ...props,
            subheading: true
          })
        },
        {
          label: 'Large',
          transformProps: ({ hero, ...props }) => ({
            ...props,
            large: true
          })
        },
        {
          label: 'Standard',
          transformProps: ({ hero, ...props }) => props
        },
        {
          label: 'Small',
          transformProps: ({ hero, ...props }) => ({
            ...props,
            small: true
          })
        }
      ]
    }
  ]
};
