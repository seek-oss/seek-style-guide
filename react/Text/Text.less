@import (reference) "~seek-style-guide/theme";

.root {
  padding-bottom: @row-height * 2;
  display: block;

  .root {
    position: relative;
    outline: 2px solid red;
    &::after {
      content: "ERROR: Multiple <Text> components nested";
      background: red;
      font-weight: @sk-bold;
      color: @sk-white;
      position: absolute;
      top: 0;
      right: 0;
      padding: 0 @gutter-width;
    }
  }
}

.standard {
  &:not(.baseline) {
    .standardTextResponsive(@baseline: false);
  }
  &.baseline {
    .standardTextResponsive(@baseline: true);
  }
  padding-bottom: @row-height;
}

.small {
  &:not(.baseline) {
    .smallTextResponsive(@baseline: false);
  }
  &.baseline {
    .smallTextResponsive(@baseline: true);
  }
  padding-bottom: @row-height;
}

.large {
  &:not(.baseline) {
    .largeTextResponsive(@baseline: false);
  }
  &.baseline {
    .largeTextResponsive(@baseline: true);
  }
  padding-bottom: @row-height;
}

.subheading {
  &:not(.baseline) {
    .subheadingTextResponsive(@baseline: false);
  }
  &.baseline {
    .subheadingTextResponsive(@baseline: true);
  }
}

.headline {
  &:not(.baseline) {
    .headlineTextResponsive(@baseline: false);
  }
  &.baseline {
    .headlineTextResponsive(@baseline: true);
  }
}

.heading {
  &:not(.baseline) {
    .headingTextResponsive(@baseline: false);
  }
  &.baseline {
    .headingTextResponsive(@baseline: true);
  }
}

.hero {
  &:not(.baseline) {
    .heroTextResponsive(@baseline: false);
  }
  &.baseline {
    .heroTextResponsive(@baseline: true);
  }
}

.light {
  font-weight: @sk-light;
}

.raw {
  padding-bottom: 0;
}

.bullet {
  display: list-item;
  list-style-type: disc;
  margin-left: 1.5 * @gutter-width;
}
