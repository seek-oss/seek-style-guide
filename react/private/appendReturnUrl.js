export default (url, returnUrl) => {
  const urlSuffix = returnUrl
    ? `?returnUrl=${encodeURIComponent(returnUrl)}`
    : '';

  return `${url}${urlSuffix}`;
};
