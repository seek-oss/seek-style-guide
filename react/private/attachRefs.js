// @flow
export default (...refs: Array<Function>) => (
  ref: ?React$ElementRef<'input'> | ?React$ElementRef<'textarea'>
): void => {
  refs.forEach(callRef => {
    if (typeof callRef === 'function') {
      callRef(ref);
    }
  });
};
