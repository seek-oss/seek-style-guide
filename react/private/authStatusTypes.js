export const AUTHENTICATED = 'authenticated';
export const UNAUTHENTICATED = 'unauthenticated';
export const AUTH_PENDING = 'pending';
