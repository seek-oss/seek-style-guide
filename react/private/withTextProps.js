import React from 'react';
import PropTypes from 'prop-types';

import omit from 'lodash/omit';
import has from 'lodash/has';
import some from 'lodash/some';
import includes from 'lodash/includes';
import forEach from 'lodash/forEach';

export const sizes = [
  'small',
  'standard',
  'large',
  'subheading',
  'heading',
  'headline',
  'hero'
];

const getBooleanSizePropTypes = () => {
  const booleanProps = {};

  forEach(sizes, size => {
    booleanProps[size] = PropTypes.bool;
  });

  return booleanProps;
};

export const SizePropTypes = {
  // eslint-disable-next-line consistent-return
  size: (props, propName, componentName) => {
    if (props.size && !includes(sizes, props.size)) {
      return new Error(
        `Invalid prop size='${props.size}' supplied to ${componentName}`
      );
    }

    if (props.size && some(sizes, size => has(props, size))) {
      return new Error(
        `Seems that you've accidentially supplied boolean size along with size='${
          props.size
        }' to ${componentName}, please remove one of them. Otherwise boolean prop will overwrite the 'size' prop.`
      );
    }
  },
  ...getBooleanSizePropTypes()
};

const parseBooleanSize = props => {
  const sizeProps = {};

  forEach(sizes, size => {
    if (props[size]) {
      sizeProps.size = size;
    }
  });

  return sizeProps;
};

const withTextProps = OriginalComponent => {
  const DecoratedComponent = props => {
    const sizeProp = parseBooleanSize(props);

    const newProps = {
      ...omit(props, sizes),
      ...sizeProp
    };

    return <OriginalComponent {...newProps} />;
  };

  DecoratedComponent.propTypes = SizePropTypes;

  DecoratedComponent.displayName = OriginalComponent.displayName;

  return DecoratedComponent;
};

export default withTextProps;
