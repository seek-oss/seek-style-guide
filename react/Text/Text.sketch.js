import React from 'react';
import mapKeys from 'lodash/mapKeys';
import lessToJs from 'less-vars-to-js';
import type from '!!raw-loader!seek-style-guide/theme/type/type.less';
import Text from './Text';
import TextLink from '../TextLink/TextLink';

const sizes = [
  'hero',
  'headline',
  'heading',
  'subheading',
  'large',
  'standard',
  'small'
];
const typeVars = lessToJs(type);
const baseFontSize = parseInt(typeVars['@base-font-size'], 10);
const symbolNames = sizes.reduce((obj, size, i) => {
  const [firstChar, ...restChars] = size;
  const name = `${firstChar.toUpperCase()}${restChars.join('')}`;

  const desktopSize =
    parseFloat(typeVars[`@${size}-type-scale`]) * baseFontSize;
  const mobileSize =
    parseFloat(typeVars[`@${size}-type-scale-mobile`]) * baseFontSize;

  const symbolName = `${i +
    1}. ${name} â€” Desktop: ${desktopSize}px, Mobile: ${mobileSize}px`;

  return { ...obj, [size]: symbolName };
}, {});

export const text = {
  [`${symbolNames.hero}`]: <Text hero>Hero text</Text>,
  [`${symbolNames.headline}`]: <Text headline>Headline text</Text>,
  [`${symbolNames.heading}`]: <Text heading>Heading text</Text>,
  [`${symbolNames.subheading}`]: <Text subheading>Subheading text</Text>,
  [`${symbolNames.large}`]: <Text large>Large text</Text>,
  [`${symbolNames.standard}/1. Default`]: <Text>Standard text</Text>,
  [`${symbolNames.standard}/2. Secondary`]: (
    <Text secondary>Standard critical text</Text>
  ),
  [`${symbolNames.standard}/3. Strong`]: (
    <Text strong>Standard strong text</Text>
  ),
  [`${symbolNames.standard}/4. Positive`]: (
    <Text positive>Standard positive text</Text>
  ),
  [`${symbolNames.standard}/5. Critical`]: (
    <Text critical>Standard critical text</Text>
  ),
  [`${symbolNames.standard}/6. Link`]: (
    <Text>
      <TextLink href="#">Standard link text</TextLink>
    </Text>
  ),
  [`${symbolNames.small}`]: <Text small>Small text</Text>
};

// Export text styles as symbols
// (at least until Sketch libraries support shared text styles)
export const symbols = mapKeys(text, (value, key) => `Text/${key}`);
