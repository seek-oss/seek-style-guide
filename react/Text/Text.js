import styles from './Text.less';
import React from 'react';
import PropTypes from 'prop-types';

import classnames from 'classnames';

import stylesPositive from '../Positive/Positive.less';
import stylesCritical from '../Critical/Critical.less';
import stylesSecondary from '../Secondary/Secondary.less';
import stylesHighlight from '../Highlight/Highlight.less';
import stylesInfo from '../Info/Info.less';
import stylesStrong from '../Strong/Strong.less';
import stylesRegular from '../Regular/Regular.less';

import withTextProps, { SizePropTypes } from '../private/withTextProps';

const Text = ({
  component,
  bullet,
  children,
  className,
  size,
  raw,
  positive,
  critical,
  secondary,
  highlight,
  info,
  strong,
  regular,
  light,
  baseline,
  ...restProps
}) => {
  const Component = component || (bullet ? 'li' : 'span');
  return (
    <Component
      {...restProps}
      className={classnames({
        [styles.root]: true,
        [className]: className,
        [styles[size]]: size,
        [styles.raw]: raw,
        [styles.bullet]: bullet,
        [styles.baseline]: baseline
      })}
    >
      <span
        className={classnames({
          [stylesPositive.root]: positive,
          [stylesCritical.root]: critical,
          [stylesSecondary.root]: secondary,
          [stylesHighlight.root]: highlight,
          [stylesHighlight.secondary]: highlight && secondary,
          [stylesInfo.root]: info,
          [stylesStrong.root]: strong,
          [stylesRegular.root]: regular,
          [styles.light]: light
        })}
      >
        {children}
      </span>
    </Component>
  );
};

Text.displayName = 'Text';

Text.propTypes = {
  component: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.string,
    PropTypes.func
  ]),
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  bullet: PropTypes.bool,
  ...SizePropTypes,
  raw: PropTypes.bool,
  positive: PropTypes.bool,
  critical: PropTypes.bool,
  secondary: PropTypes.bool,
  highlight: PropTypes.bool,
  info: PropTypes.bool,
  strong: PropTypes.bool,
  regular: PropTypes.bool,
  light: PropTypes.bool,
  baseline: PropTypes.bool
};

Text.defaultProps = {
  baseline: true,
  bullet: false,
  size: 'standard'
};

export default withTextProps(Text);
