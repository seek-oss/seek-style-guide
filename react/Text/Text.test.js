import { render } from 'enzyme';
import React from 'react';
import Text from './Text';

import { sizes } from '../private/withTextProps';

describe('Text', () => {
  it('should render with defaults', () => {
    expect(render(<Text>Hello</Text>)).toMatchSnapshot();
  });
  it('should render with a custom component', () => {
    expect(render(<Text component="div">Hello</Text>)).toMatchSnapshot();
  });
  it('should render a bullet point', () => {
    expect(render(<Text bullet>Hello</Text>)).toMatchSnapshot();
  });

  describe('sizes', () => {
    sizes.forEach(size => {
      it(`should render as ${size}`, () => {
        expect(
          render(<Text {...{ [size]: true }}>Hello</Text>)
        ).toMatchSnapshot();
        expect(render(<Text size={size}>Hello</Text>)).toMatchSnapshot();
      });
    });
  });
});
