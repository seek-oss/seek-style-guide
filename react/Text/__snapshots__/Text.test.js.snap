// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should render a bullet point 1`] = `
<li
  class="Text__root Text__standard Text__bullet Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</li>
`;

exports[`Text should render with a custom component 1`] = `
<div
  class="Text__root Text__standard Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</div>
`;

exports[`Text should render with defaults 1`] = `
<span
  class="Text__root Text__standard Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as heading 1`] = `
<span
  class="Text__root Text__heading Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as heading 2`] = `
<span
  class="Text__root Text__heading Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as headline 1`] = `
<span
  class="Text__root Text__headline Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as headline 2`] = `
<span
  class="Text__root Text__headline Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as hero 1`] = `
<span
  class="Text__root Text__hero Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as hero 2`] = `
<span
  class="Text__root Text__hero Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as large 1`] = `
<span
  class="Text__root Text__large Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as large 2`] = `
<span
  class="Text__root Text__large Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as small 1`] = `
<span
  class="Text__root Text__small Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as small 2`] = `
<span
  class="Text__root Text__small Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as standard 1`] = `
<span
  class="Text__root Text__standard Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as standard 2`] = `
<span
  class="Text__root Text__standard Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as subheading 1`] = `
<span
  class="Text__root Text__subheading Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;

exports[`Text sizes should render as subheading 2`] = `
<span
  class="Text__root Text__subheading Text__baseline"
>
  <span
    class=""
  >
    Hello
  </span>
</span>
`;
